src/librustc/traits/specialize/mod.rs

```diff
@@ -156,6 +156,11 @@ pub fn specializes<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
         return r;
     }
 
+    if tcx.impl_always_allowed_to_overlap(impl1_def_id)
+        && tcx.impl_always_allowed_to_overlap(impl2_def_id) {
+        return true;
+    }
```

Comment:
I think I would prefer not to modify `specializes` to return true. It makes the specialization graph "cyclic". I think we can make a more direct modification to `select.rs` so that where it calls specialized, it would do this instead:

```rust
return impls_are_allowed_to_ovelap(other_def, victim_def) ||
       traits::specializes(tcx, other_def, victim_def);
```

This seems to more directly express what is going on.